wdl-format/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+                    let comment = Comment::cast(
+                        sibling
+                            .as_token()
+                            .expect("Comment should be a token")
+                            .clone(),
+                    )
+                    .expect("Comment should cast to a comment");
+                    comment.format(comment_buffer, state)?;
+                    // If the comment does not start with "##", prepend a '#' to it
+                    if !comment_buffer.starts_with("##") && !comment_buffer.starts_with("#@") {
+                        comment_buffer.insert(0, '#');
+                    }
+                    if comment_buffer.starts_with("#@") {
+                        lint_directives.push(comment_buffer.clone());
+                    } else {
+                        preamble_comments.push(comment_buffer.clone());
+                    }
+                    comment_buffer.clear();
+                }
+                SyntaxKind::Whitespace => {
+                    // Ignore
+                }
+                _ => {
+                    unreachable!("Unexpected syntax kind: {:?}", sibling.kind());
+                }
+            }
+        }
+
+        for comment in preamble_comments.iter().rev() {
+            write!(writer, "{}", comment)?;
+        }
+
+        // If there are preamble comments, ensure a blank line is inserted
+        if !preamble_comments.is_empty() {
+            write!(writer, "{}", NEWLINE)?;
+        }
+
+        for comment in lint_directives.iter().rev() {
+            write!(writer, "{}", comment)?;
+        }
+
+        let version_keyword = first_child_of_kind(self.syntax(), SyntaxKind::VersionKeyword);
+        write!(writer, "{}", version_keyword)?;
+        format_inline_comment(&version_keyword, writer, state, true)?;
+
+        let version = self.version();
+
+        format_preceding_comments(
+            &SyntaxElement::from(version.syntax().clone()),
+            writer,
+            state,
+            true,
+        )?;
+        state.space_or_indent(writer)?;
+        version.format(writer, state)?;
+        format_inline_comment(
+            &SyntaxElement::from(self.syntax().clone()),
+            writer,
+            state,
+            false,
+        )
+    }
+}
+
+impl Formattable for Ident {
+    fn format<T: std::fmt::Write>(&self, writer: &mut T, _state: &mut State) -> std::fmt::Result {
+        write!(writer, "{}", self.as_str())
+    }
+}
+
+/// Find an expected child element of the specified kind.
+///
+/// # Panics
+/// Panics if the child element is not found.
+pub fn first_child_of_kind(node: &SyntaxNode, kind: SyntaxKind) -> SyntaxElement {
+    node.children_with_tokens()
+        .find(|element| element.kind() == kind)
+        .unwrap_or_else(|| panic!("Expected to find a child of kind: {kind:?}"))
+}
+
+/// Format a WDL document.
+pub fn format_document(code: &str) -> Result<String, Vec<Diagnostic>> {
```

Comment:
> * First, this needs to be recursive all of the way down. What I mean by that is that `Document` should implement `Formattable`, and you should be able to call `Document::format` as your entrypoint (or `format` on any other `SyntaxKind` if you just want to format a subset of the file).

That's done, but is complicated by the trait of `Formattable` that the `format()` signature takes a `state: State`. This isn't very ergonomic and requires a `let mut state = State::default()` statement before any call `format()`. The `State` requirement is so that an element knows how far it needs to be indented, but that really only applies to elements which can be nested (anything in a metadata section or an `Expr`, and `WorkflowItem`).

I'm wondering if there should be two different `Formattable` traits, one which needs a `state` (`NestedFormattable`) and one which doesn't (`FlatFormattable`?). Because many elements can only appear at certain indent levels. There's top level `DocumentItem` which must be `indent_level = 0` and then `TaskItem` and `WorkflowItem` that must be `indent_level = 1` etc.

> * I think it will be up to the user as to how they'd like to collect the output. Here, you've forced them to allocate a `String` on each format, but they may not want to do that.

👍 

> * You've also used the AST nodes underneath document. I think this should only require working at the `SyntaxKind` level to avoid all of the casting everywhere (and, logically, this is probably the place that most closely matches what is "formattable").

I'm not sure I'm understanding you on this point. Are you suggesting that `format` only operate on the CST level, and ignores the AST level? I think this would actually greatly complicate things, particularly when we want to move things around and group them. Operating at the AST level lets me cherrypick  elements in the order in which we want them in the output. Roughly only visiting things once. Operating at the CST level is going to require traversing up and down the CST multiple times.

> * The above also allows you to not have to do the document parsing yourself, which is good; we don't really care about how the `Document` was created here (it could even be hand crafted programmatically using some future API).

I believe this is already handled by the existing code? Most every AST node (with some exceptions) has implemented `Formattable`, allowing them to be called "individually" or outside the context of a Document.

Reply:
Another thought: currently every `format()` call assumes the element is well-formed and has passed both CST and AST validation. Of course eventually we want `format()` to be infallible so it can still format problematic documents (that are perhaps being edited in VScode). But we aren't there in this PR. So that's a known limitation of the current implementation.